FPGA: Fill out docstrings for LVDS,FIFO Bringup

jwenner opened the pull request:
Fill out docstrings for functions used in DAC LVDS and FIFO bringup.

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -677,8 +677,25 @@ def _setPolarity(self, chan, invert):
     def _checkPHOF(self, op, fifoReadings, counterValue):
-        # Determine for which PHOFs (FIFO offsets) the FIFO counter equals
-        # counterValue.
+        """ Determine correct PHOF for FIFO counter of counterValue
```
To be completely 100% correct, I believe the standard is to not have a after the `"""` and to end this line with a period.

jwenner replied:
Point taken on the period. However, for the space, in the pyle contributing file we have a space between the `"""` and the start of the text. What do I do?

jwenner replied:
Of course, it helps if I'm consistent (cf setLVDS vs setFIFO). Suggestions?

pomalley replied:
you probably saw, but contributing.md got updated w.r.t. this.

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -677,8 +677,25 @@ def _setPolarity(self, chan, invert):
+        """ Determine correct PHOF for FIFO counter of counterValue
+
+        If finds such a PHOF, sets PHOF and reads back FIFO counter to
+        verify that FIFO counter equals counterValue.
```
"If finds..."? Is this supposed to be "If we find a valid PHOF, set PHOF and..."

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -677,8 +677,25 @@ def _setPolarity(self, chan, invert):
+        verify that FIFO counter equals counterValue.
+
+        If no PHOF can be found to get an acceptable FIFO counter, PHOF will
+        be returned as -1, and success will be False.
```
Generally try to keep things in the imperative. I suggest: "If no PHOF can be found to get an acceptable FIFO counter, return (-1, False)."

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -677,8 +677,25 @@ def _setPolarity(self, chan, invert):
+                         for FIFO counter to be counterValue and, after
+                         setting PHOF, that FIFO counter is counterValue
+        :rtype: (int, bool)
+        """
```
I think it would be better to define PHOF (as phase offset? FIFO phase offset?) when it's first introduced, after the first line. Since we already know that the goal is to find the valid PHOF, I would condense the :return: part to:
```
:return: (PHOF for given counterValue, success)
```

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -765,21 +782,33 @@ def setPolarity(self, chan, invert):
+                            A=1: An LVDS input above the IEEE input
+                                 voltage specification limit
+                            B=1: An LVDS input below the IEEE input
+                                 voltage specification limit
```
"An LVDS input [was?] above/below..." ?

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -765,21 +782,33 @@ def setPolarity(self, chan, invert):
+                                 voltage specification limit
+                            B=1: An LVDS input below the IEEE input
+                                 voltage specification limit
+                            C=1: Sampling in correct data cycle
```
What does this mean?

jwenner replied:
Uh, er, well...

To quote the data sheet:
```
0, phase measurement—sampling in the previous or following DATA cycle.
1, phase measurement—sampling in the correct DATA cycle.
```
Perhaps referring to one of the DATA clocks? But which one (maybe DATACLK_IN, but not sure).

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -765,21 +782,33 @@ def setPolarity(self, chan, invert):
+                            C=1: Sampling in correct data cycle
+        :rtype: (bool, int, int, int,
+                 tuple(list[int], list[bool], list[bool]),
+                 int)
```
It is unfortunate that we have such a convoluted return type...

jwenner replied:
Agreed. If I were doing this again, I'd return (success, t, checkHex) and axe the MSD/MHD/cluster garbage since these don't seem to be useful. These come from when the bringup routine actually plotted where MSD and MHD flipped.

However, changing this would require changing not only the server but also the bringup routine & possibly ScanPhases code. I think this is out of scope for this PR. But certainly this should be done at some point for greater clarity in the return (and to clean up that wonderful dac_bringup return tag).

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -850,6 +879,27 @@ def func():
+        :param int op: 2 for DAC A, 3 for DAC B
+        :param int targetFifo: Desired targetFifo. If None, use value
+                               from board registry entry.
+        :return: found - If procedure successfull
```
`sucessful`

pomalley commented on GHzDACs/Cleanup/dac.py:
```diff
@@ -850,6 +879,27 @@ def func():
+                 targetFifo - Number bits between read/write signals
+                              for 8-bit buffer.
+        :rtype: (bool, bool, int, int, int)
+        """
```
Generally the same comments for this one as for _checkPHOF.

pomalley commented on GHzDACs/ghz_fpga_server.py:
```diff
@@ -1933,14 +1933,27 @@ def dac_init(self, c, chan, signed=False):
+                 checkHex - In binary, '0bABC', where
+                            A=1: An LVDS input above specification
+                            B=1: An LVDS input below specification
+                            C=1: Sampling in correct data cycle
```
ditto for here as in setLVDS

pomalley commented:
Thanks for doing this Jim, never had a clue what was going on here. I had some comments so let me know what you think of them.

Also, in general it looks like the lines don't go out to 80 characters--is this the case?

maffoo commented:
Question for y'all: `dac_lvds` is a labrad setting, so its doc string gets exposed to the world as labrad documentation. The python conventions for documentation may look strange, especially given that this can be accessed from other languages that do things differently. Not saying we should change what @jwenner has done, but just curious what you all think about how to handle this generally.

pomalley commented:
I have  no idea. :-)

maffoo commented:
that was my answer too. glad we're in agreement :)

pomalley commented:
I suppose we should go with the Google python style guide because that is very readable, and keep it to 80 chars because people read it in the console. Other than that... depends on what documentation generator we use, if any?

And you can provide some sort of annotations for the parameters to the setting, right? But IIRC there's some reason we were moving away from that.

maffoo commented:
The issue is we're using the same docstring for two purposes: (1) documenting the python method, for people reading the python code, and (2) documenting the labrad setting for people calling it from labrad. Some of the info needed for (1) is not relevant for (2), and exactly what is needed for (2) depends somewhat on which language api one is using to call it. But anyway, I don't want to hold up this PR because this is a larger discussion.

jwenner commented:
I was trying to follow PEP8 on the line length. In particular,
`
For flowing long blocks of text with fewer structural restrictions (docstrings or comments), the line length should be limited to 72 characters. 
`
Should I instead be using 80 characters as the limit?

pomalley commented:
I think limiting comments to 72 characters is silly. PEP8 doesn't really explain why so I'm inclined to stick with 80.

jwenner commented:
@pomalley, could you please take a look now? Tried to implement the changes you suggested. Also added additional explanation on the outputs of the setFIFO + dac_fifo functions. Also converted docstrings in ghz_fpga_server to (I believe) Google format. Thanks for the review!

Since they're hidden as outdated diffs, note my comments above about the LVDS data cycle and return type.

pomalley commented:
LGTM
